File: app/src/main/java/wangdaye/com/geometricweather/ui/widget/weatherView/materialWeatherView/implementor/SunImplementor.java
package wangdaye.com.geometricweather.ui.widget.weatherView.materialWeatherView.implementor;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;

import androidx.annotation.ColorInt;
import androidx.annotation.Size;

import wangdaye.com.geometricweather.ui.widget.weatherView.materialWeatherView.MaterialWeatherView;

/**
 * Clear day implementor.
 * */

public class SunImplementor extends MaterialWeatherView.WeatherAnimationImplementor {

    private Paint paint;
    private float[] angles;
    private float[] unitSizes;

    public SunImplementor(@Size(2) int[] canvasSizes) {
        this.paint = new Paint();
        paint.setStyle(Paint.Style.FILL);
        paint.setAntiAlias(true);
        paint.setColor(Color.rgb(253, 84, 17));

        this.angles = new float[3];

        this.unitSizes = new float[3];
        unitSizes[0] = (float) (0.5 * 0.47 * canvasSizes[0]);
        unitSizes[1] = (float) (1.7794 * unitSizes[0]);
        unitSizes[2] = (float) (3.0594 * unitSizes[0]);
    }

    @Override
    public void updateData(@Size(2) int[] canvasSizes, long interval,
                           float rotation2D, float rotation3D) {
        for (int i = 0; i < angles.length; i ++) {
            angles[i] = (float) ((angles[i] + (90.0 / (3000 + 1000 * i) * interval)) % 90);
        }
    }

    @Override
    public void draw(@Size(2) int[] canvasSizes, Canvas canvas,
                     float displayRate, float scrollRate, float rotation2D, float rotation3D) {

        canvas.drawColor(Color.argb((int) (displayRate * 255), 253, 188, 76));

        if (scrollRate < 1) {
            float deltaX = (float) (Math.sin(rotation2D * Math.PI / 180.0) * 0.3 * canvasSizes[0]);
            float deltaY = (float) (Math.sin(rotation3D * Math.PI / 180.0) * -0.3 * canvasSizes[0]);

            canvas.translate(
                    canvasSizes[0] + deltaX,
                    (float) (0.0333 * canvasSizes[0] + deltaY));

            paint.setAlpha((int) (displayRate * (1 - scrollRate) * 255 * 0.40));
            canvas.rotate(angles[0]);
            for (int i = 0; i < 4; i ++) {
                canvas.drawRect(-unitSizes[0], -unitSizes[0], unitSizes[0], unitSizes[0], paint);
                canvas.rotate(22.5F);
            }
            canvas.rotate(-90 - angles[0]);

            paint.setAlpha((int) (displayRate * (1 - scrollRate) * 255 * 0.16));
            canvas.rotate(angles[1]);
            for (int i = 0; i < 4; i ++) {
                canvas.drawRect(-unitSizes[1], -unitSizes[1], unitSizes[1], unitSizes[1], paint);
                canvas.rotate(22.5F);
            }
            canvas.rotate(-90 - angles[1]);

            paint.setAlpha((int) (displayRate * (1 - scrollRate) * 255 * 0.08));
            canvas.rotate(angles[2]);
            for (int i = 0; i < 4; i ++) {
                canvas.drawRect(-unitSizes[2], -unitSizes[2], unitSizes[2], unitSizes[2], paint);
                canvas.rotate(22.5F);
            }
            canvas.rotate(-90 - angles[2]);
        }
    }

    @ColorInt
    public static int getThemeColor() {
        return Color.rgb(253, 188, 76);
    }
}

File: ui-weather-view/src/main/java/org/breezyweather/ui/theme/weatherView/materialWeatherView/implementor/SunImplementor.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.ui.theme.weatherView.materialWeatherView.implementor

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import androidx.annotation.ColorInt
import androidx.annotation.Size
import org.breezyweather.ui.theme.weatherView.materialWeatherView.MaterialWeatherView.WeatherAnimationImplementor
import kotlin.math.sin

/**
 * Clear day implementor.
 */
class SunImplementor(
    @Size(2) canvasSizes: IntArray,
    animate: Boolean,
) : WeatherAnimationImplementor() {
    private val mAnimate = animate
    private val mPaint = Paint().apply {
        style = Paint.Style.FILL
        isAntiAlias = true
        color = Color.rgb(253, 84, 17)
    }
    private val mAngles = FloatArray(3)
    private val mUnitSizes: FloatArray = floatArrayOf(
        (0.5 * 0.47 * canvasSizes[0]).toFloat(),
        (1.7794 * 0.5 * 0.47 * canvasSizes[0]).toFloat(),
        (3.0594 * 0.5 * 0.47 * canvasSizes[0]).toFloat()
    )

    override fun updateData(
        @Size(2) canvasSizes: IntArray,
        interval: Long,
        rotation2D: Float,
        rotation3D: Float,
    ) {
        for (i in mAngles.indices) {
            mAngles[i] = ((mAngles[i] + 90.0 / (3000 + 1000 * i) * interval) % 90).toFloat()
        }
    }

    override fun draw(
        @Size(2) canvasSizes: IntArray,
        canvas: Canvas,
        scrollRate: Float,
        rotation2D: Float,
        rotation3D: Float,
    ) {
        if (scrollRate < 1) {
            val deltaX = (sin(rotation2D * Math.PI / 180.0) * 0.3 * canvasSizes[0]).toFloat()
            val deltaY = (sin(rotation3D * Math.PI / 180.0) * -0.3 * canvasSizes[0]).toFloat()
            canvas.translate(canvasSizes[0] + deltaX, (SUN_POSITION * canvasSizes[0] + deltaY).toFloat())

            arrayOf(SMALL_SUN_ALPHA, MEDIUM_SUN_ALPHA, LARGE_SUN_ALPHA).forEachIndexed { index, alpha ->
                mPaint.alpha = ((1 - scrollRate) * 255 * alpha).toInt()
                canvas.rotate(mAngles[index])
                for (i in 0..3) {
                    canvas.drawRect(
                        -mUnitSizes[index],
                        -mUnitSizes[index],
                        mUnitSizes[index],
                        mUnitSizes[index],
                        mPaint
                    )
                    canvas.rotate(22.5f)
                }
                canvas.rotate(-90 - mAngles[index])
            }
        }
    }

    companion object {
        const val SUN_POSITION = 0.0333
        const val SMALL_SUN_ALPHA = 0.40
        const val MEDIUM_SUN_ALPHA = 0.16
        const val LARGE_SUN_ALPHA = 0.08

        @get:ColorInt
        val themeColor: Int
            get() = Color.rgb(253, 188, 76)
    }
}
